File: backend/src/prechecks/concrete/node/jvm-heap-settings.precheck.ts
import { ElasticClient } from "../../../clients/elastic.client";
import { ClusterNodeType } from "../../../models/cluster-node.model";
import { BaseNodePrecheck } from "../../base/base-node-precheck";
import { ExecutionMode, PrecheckType } from "../../types/enums";
import { PrecheckExecutionRequest, NodeContext } from "../../types/interfaces";

export class JvmHeapSettingsPrecheck extends BaseNodePrecheck {
	constructor() {
		super(
			{
				id: "elasticsearch_jvm_heap_check",
				name: "JVM heap settings check",
				type: PrecheckType.NODE,
				mode: ExecutionMode.CODE,
			},
			ClusterNodeType.ELASTIC
		);
	}

	protected async runForContext(request: PrecheckExecutionRequest<NodeContext>): Promise<void> {
		const client = await ElasticClient.buildClient(request.context.node.clusterId);
		const es = client.getClient();

		const nodeId = request.context.node.nodeId;

		const nodesInfo = await es.nodes.info({ node_id: nodeId, metric: "jvm" });

		const node = nodesInfo.nodes[nodeId];
		if (!node) {
			throw new Error(`Node with ID ${nodeId} not found`);
		}

		const name = node.name;
		const heapInit = node.jvm?.mem.heap_init_in_bytes;
		const heapMax = node.jvm?.mem.heap_max_in_bytes;

		if (heapInit && heapMax) {
			const heapInitGB = heapInit / 1024 ** 3;
			const heapMaxGB = heapMax / 1024 ** 3;

			this.addLog(request, `${name}: -Xms=${heapInitGB}GB, -Xmx=${heapMaxGB}GB`);

			if (heapInit !== heapMax) {
				throw new Error(`${name} has mismatched -Xms and -Xmx. Must be equal.`);
			}

			// if (heapMaxGB < 4 || heapMaxGB > 32) {
			// 	throw new Error(`${name} has suspicious heap size: ${heapMaxGB}GB. Recommended: 4–32 GB.`);
			// }
		} else {
			const reason =
				!heapInit && !heapMax
					? `heap_init and heap_max are both missing`
					: !heapInit
						? `heap_init is missing`
						: `heap_max is missing`;

			await this.addLog(request, `${name}: Skipping JVM heap check — ${reason}.`);
		}
	}
}

File: server/src/main/java/co/hyperflex/prechecks/concrete/node/elastic/JvmHeapSettingsPrecheck.java
package co.hyperflex.prechecks.concrete.node.elastic;

import co.elastic.clients.elasticsearch.ElasticsearchClient;
import co.elastic.clients.elasticsearch.nodes.NodesInfoResponse;
import co.elastic.clients.elasticsearch.nodes.info.NodeInfo;
import co.elastic.clients.elasticsearch.nodes.info.NodeInfoJvmMemory;
import co.elastic.clients.elasticsearch.nodes.info.NodeJvmInfo;
import co.hyperflex.entities.precheck.PrecheckSeverity;
import co.hyperflex.prechecks.contexts.NodeContext;
import co.hyperflex.prechecks.core.BaseElasticNodePrecheck;
import java.io.IOException;
import java.util.Map;
import org.slf4j.Logger;
import org.springframework.stereotype.Component;

@Component
public class JvmHeapSettingsPrecheck extends BaseElasticNodePrecheck {

  @Override
  public String getName() {
    return "JVM heap settings check";
  }

  @Override
  public PrecheckSeverity getSeverity() {
    return PrecheckSeverity.WARNING;
  }

  @Override
  public void run(NodeContext context) {
    String nodeId = context.getNode().getId();
    ElasticsearchClient client = context.getElasticClient().getElasticsearchClient();
    Logger logger = context.getLogger();

    try {
      NodesInfoResponse response = client.nodes().info(r -> r.nodeId(nodeId).metric("jvm"));

      Map<String, NodeInfo> nodes = response.nodes();
      NodeInfo node = nodes.get(nodeId);

      if (node == null) {
        throw new RuntimeException("Node with ID [" + nodeId + "] not found");
      }

      String name = node.name();
      NodeJvmInfo jvmStats = node.jvm();
      if (jvmStats == null || jvmStats.mem() == null) {
        logger.info("{}: Skipping JVM heap check — JVM memory stats missing.", name);
        return;
      }

      NodeInfoJvmMemory memStats = jvmStats.mem();
      long heapInit = memStats.heapInitInBytes();
      long heapMax = memStats.heapMaxInBytes();

      double heapInitGB = heapInit / Math.pow(1024, 3);
      double heapMaxGB = heapMax / Math.pow(1024, 3);

      logger.info("{}: -Xms={}GB, -Xmx={}GB", name, heapInitGB, heapMaxGB);

      if (heapInit != heapMax) {
        logger.error("{} has mismatched -Xms and -Xmx values. They must be equal.", name);
        throw new RuntimeException();
      }

    } catch (IOException e) {
      logger.error("Failed to fetch JVM info for node: {}", nodeId, e);
      throw new RuntimeException("Failed to fetch JVM info for node: " + nodeId, e);
    }
  }
}
